File: src/overlay_sav.py
#!/usr/bin/env python3
"""
Overlay / merge data file (terutama .sav SPSS) dengan opsi prioritas kolom.

Contoh:
  python overlay_sav.py \
    --base data/base.sav \
    --overlay data/patch.sav \
    --keys id \
    --how left \
    --method replace \
    --output out/hasil.sav \
    --report out/report.json
"""

from __future__ import annotations

import argparse
import json
import os
from dataclasses import dataclass
from typing import Iterable


def _require(name: str):
    try:
        return __import__(name)
    except ImportError as e:
        raise SystemExit(
            f"Dependency '{name}' belum terpasang. Jalankan: pip install -r requirements-python.txt"
        ) from e


pd = _require("pandas")
pyreadstat = _require("pyreadstat")


@dataclass
class IOResult:
    df: "pd.DataFrame"
    source_type: str


def parse_csv_list(raw: str | None) -> list[str]:
    if not raw:
        return []
    return [x.strip() for x in raw.split(",") if x.strip()]


def ext(path: str) -> str:
    return os.path.splitext(path)[1].lower()


def read_table(path: str) -> IOResult:
    ex = ext(path)
    if ex == ".sav":
        df, _meta = pyreadstat.read_sav(path)
        return IOResult(df=df, source_type="sav")
    if ex == ".csv":
        return IOResult(df=pd.read_csv(path), source_type="csv")
    if ex in {".xlsx", ".xls"}:
        return IOResult(df=pd.read_excel(path), source_type="excel")
    if ex == ".parquet":
        return IOResult(df=pd.read_parquet(path), source_type="parquet")
    raise SystemExit(f"Format input tidak didukung: {path}")


def write_table(df: "pd.DataFrame", path: str) -> None:
    os.makedirs(os.path.dirname(path) or ".", exist_ok=True)
    ex = ext(path)
    if ex == ".sav":
        pyreadstat.write_sav(df, path)
        return
    if ex == ".csv":
        df.to_csv(path, index=False)
        return
    if ex in {".xlsx", ".xls"}:
        df.to_excel(path, index=False)
        return
    if ex == ".parquet":
        df.to_parquet(path, index=False)
        return
    raise SystemExit(f"Format output tidak didukung: {path}")


def normalize_columns(df: "pd.DataFrame") -> "pd.DataFrame":
    out = df.copy()
    out.columns = [str(c).strip().lower().replace(" ", "_") for c in out.columns]
    return out


def ensure_keys(df: "pd.DataFrame", keys: Iterable[str], label: str) -> None:
    missing = [k for k in keys if k not in df.columns]
    if missing:
        raise SystemExit(f"Kolom key tidak ditemukan di {label}: {missing}")


def _normalize_key_value(x):
    if pd.isna(x):
        return pd.NA

    s = str(x).strip()
    if s == "" or s.lower() in {"nan", "none", "<na>"}:
        return pd.NA

    # Jika string angka berakhiran .0 (mis. 92.0), ubah jadi 92
    if s.replace(".", "", 1).isdigit() and s.endswith(".0"):
        try:
            f = float(s)
            if f.is_integer():
                return str(int(f))
        except Exception:
            pass

    return s


def normalize_key_columns(df: "pd.DataFrame", keys: list[str]) -> "pd.DataFrame":
    out = df.copy()
    for k in keys:
        if k in out.columns:
            out[k] = out[k].map(_normalize_key_value)
    return out


def overlay_merge(
    base_df: "pd.DataFrame",
    overlay_df: "pd.DataFrame",
    keys: list[str],
    how: str,
    method: str,
    include_cols: list[str],
    exclude_cols: list[str],
    clean_keys: bool = False,
) -> tuple["pd.DataFrame", dict]:
    if clean_keys:
        base_df = normalize_key_columns(base_df, keys)
        overlay_df = normalize_key_columns(overlay_df, keys)

    base_non_keys = [c for c in base_df.columns if c not in keys]

    if include_cols:
        overlay_cols = [c for c in include_cols if c in overlay_df.columns and c not in keys]
    else:
        overlay_cols = [c for c in overlay_df.columns if c not in keys]

    if exclude_cols:
        overlay_cols = [c for c in overlay_cols if c not in set(exclude_cols)]

    if not overlay_cols:
        raise SystemExit("Tidak ada kolom overlay yang dipilih. Cek include/exclude.")

    # Hindari duplikasi key di sisi overlay (ambil data terakhir)
    overlay_uniq = overlay_df[keys + overlay_cols].drop_duplicates(subset=keys, keep="last")

    merged = base_df.merge(
        overlay_uniq,
        on=keys,
        how=how,
        suffixes=("_base", "_ovr"),
        indicator=True,
    )

    overlap_cols = [c for c in overlay_cols if c in base_non_keys]

    for col in overlap_cols:
        c_base = f"{col}_base"
        c_ovr = f"{col}_ovr"
        if method == "coalesce":
            merged[col] = merged[c_base].combine_first(merged[c_ovr])
        elif method == "replace":
            merged[col] = merged[c_ovr].combine_first(merged[c_base])
        elif method == "keep_base":
            merged[col] = merged[c_base]
        elif method == "keep_overlay":
            merged[col] = merged[c_ovr]
        else:
            raise SystemExit(f"Method tidak dikenal: {method}")

    # Rapikan kolom non-overlap yang masih bersufiks
    rename_back = {}
    for c in merged.columns:
        if c.endswith("_base"):
            base_name = c[:-5]
            if base_name not in overlap_cols:
                rename_back[c] = base_name
        if c.endswith("_ovr"):
            ovr_name = c[:-4]
            if ovr_name not in overlap_cols and ovr_name not in base_df.columns:
                rename_back[c] = ovr_name

    merged = merged.rename(columns=rename_back)

    drop_cols = [f"{c}_base" for c in overlap_cols] + [f"{c}_ovr" for c in overlap_cols]
    merged = merged.drop(columns=[c for c in drop_cols if c in merged.columns])

    merge_counts = merged["_merge"].value_counts(dropna=False).to_dict()
    report = {
        "base_rows": int(len(base_df)),
        "overlay_rows": int(len(overlay_df)),
        "overlay_rows_after_dedup": int(len(overlay_uniq)),
        "output_rows": int(len(merged)),
        "merge_counts": {k: int(v) for k, v in merge_counts.items()},
        "keys": keys,
        "how": how,
        "method": method,
        "clean_keys": bool(clean_keys),
        "overlay_columns_used": overlay_cols,
        "overlap_columns": overlap_cols,
    }

    merged = merged.drop(columns=["_merge"])
    return merged, report


def build_parser() -> argparse.ArgumentParser:
    p = argparse.ArgumentParser(description="Overlay/merge file .sav/.csv/.xlsx/.parquet")
    p.add_argument("--base", required=True, help="File base (mis. data_awal.sav)")
    p.add_argument("--overlay", required=True, help="File overlay/patch")
    p.add_argument("--keys", required=True, help="Kolom key, pisahkan koma. Contoh: id atau id,tanggal")
    p.add_argument("--output", required=True, help="Output file (.sav/.csv/.xlsx/.parquet)")
    p.add_argument("--report", help="Path JSON report ringkas")
    p.add_argument("--how", default="left", choices=["left", "inner", "right", "outer"])
    p.add_argument(
        "--method",
        default="coalesce",
        choices=["coalesce", "replace", "keep_base", "keep_overlay"],
        help=(
            "Strategi untuk kolom nama sama. "
            "coalesce=base dulu, replace=overlay dulu, keep_base=tetap base, keep_overlay=pakai overlay"
        ),
    )
    p.add_argument("--include-cols", help="Kolom overlay yang dipakai (koma). Default: semua")
    p.add_argument("--exclude-cols", help="Kolom overlay yang dibuang (koma)")
    p.add_argument(
        "--normalize-cols",
        action="store_true",
        help="Normalisasi nama kolom jadi lowercase + underscore",
    )
    p.add_argument(
        "--clean-keys",
        action="store_true",
        help="Bersihkan nilai key sebelum merge (trim spasi, rapikan angka seperti 92.0 -> 92)",
    )
    return p


def main() -> None:
    args = build_parser().parse_args()

    keys = parse_csv_list(args.keys)
    include_cols = parse_csv_list(args.include_cols)
    exclude_cols = parse_csv_list(args.exclude_cols)

    if not keys:
        raise SystemExit("--keys wajib diisi")

    base_in = read_table(args.base)
    ovr_in = read_table(args.overlay)

    base_df = base_in.df
    overlay_df = ovr_in.df

    if args.normalize_cols:
        base_df = normalize_columns(base_df)
        overlay_df = normalize_columns(overlay_df)
        keys = [k.strip().lower().replace(" ", "_") for k in keys]
        include_cols = [c.strip().lower().replace(" ", "_") for c in include_cols]
        exclude_cols = [c.strip().lower().replace(" ", "_") for c in exclude_cols]

    ensure_keys(base_df, keys, "base")
    ensure_keys(overlay_df, keys, "overlay")

    result_df, report = overlay_merge(
        base_df=base_df,
        overlay_df=overlay_df,
        keys=keys,
        how=args.how,
        method=args.method,
        include_cols=include_cols,
        exclude_cols=exclude_cols,
        clean_keys=args.clean_keys,
    )

    write_table(result_df, args.output)

    if args.report:
        os.makedirs(os.path.dirname(args.report) or ".", exist_ok=True)
        with open(args.report, "w", encoding="utf-8") as f:
            json.dump(report, f, ensure_ascii=False, indent=2)

    print("[OK] Overlay selesai")
    print(json.dumps(report, ensure_ascii=False, indent=2))


if __name__ == "__main__":
    main()

File: src/overlay_sav.R
#!/usr/bin/env Rscript

suppressPackageStartupMessages({
  library(optparse)
  library(dplyr)
  library(haven)
  library(readr)
  library(jsonlite)
})

parse_csv_list <- function(x) {
  if (is.null(x) || x == "") return(character(0))
  parts <- strsplit(x, ",", fixed = TRUE)[[1]]
  trimws(parts[nzchar(trimws(parts))])
}

get_ext <- function(path) {
  tolower(tools::file_ext(path))
}

read_table <- function(path) {
  ex <- get_ext(path)
  if (ex == "sav") return(read_sav(path))
  if (ex == "csv") return(read_csv(path, show_col_types = FALSE))
  stop(sprintf("Format input tidak didukung: %s", path))
}

write_table <- function(df, path) {
  dir.create(dirname(path), recursive = TRUE, showWarnings = FALSE)
  ex <- get_ext(path)
  if (ex == "sav") {
    write_sav(df, path)
    return(invisible(NULL))
  }
  if (ex == "csv") {
    write_csv(df, path)
    return(invisible(NULL))
  }
  stop(sprintf("Format output tidak didukung: %s", path))
}

normalize_columns <- function(df) {
  names(df) <- gsub(" +", "_", tolower(trimws(names(df))))
  df
}

check_keys <- function(df, keys, label) {
  missing <- setdiff(keys, names(df))
  if (length(missing) > 0) {
    stop(sprintf("Kolom key tidak ditemukan di %s: %s", label, paste(missing, collapse = ", ")))
  }
}

option_list <- list(
  make_option("--base", type = "character", help = "File base (.sav/.csv)", metavar = "FILE"),
  make_option("--overlay", type = "character", help = "File overlay (.sav/.csv)", metavar = "FILE"),
  make_option("--keys", type = "character", help = "Kolom key (koma), contoh: id atau id,tanggal"),
  make_option("--output", type = "character", help = "Output (.sav/.csv)", metavar = "FILE"),
  make_option("--report", type = "character", default = NULL, help = "Output JSON report"),
  make_option("--how", type = "character", default = "left", help = "left|inner|right|full [default %default]"),
  make_option("--method", type = "character", default = "coalesce", help = "coalesce|replace|keep_base|keep_overlay [default %default]"),
  make_option("--include-cols", type = "character", default = NULL, help = "Kolom overlay yang dipakai (koma). Default semua"),
  make_option("--exclude-cols", type = "character", default = NULL, help = "Kolom overlay yang dibuang (koma)"),
  make_option("--normalize-cols", action = "store_true", default = FALSE, help = "Normalisasi nama kolom")
)

parser <- OptionParser(option_list = option_list)
opt <- parse_args(parser)

required <- c("base", "overlay", "keys", "output")
for (r in required) {
  if (is.null(opt[[r]]) || opt[[r]] == "") {
    print_help(parser)
    stop(sprintf("Argumen --%s wajib diisi", r))
  }
}

keys <- parse_csv_list(opt$keys)
include_cols <- parse_csv_list(opt$`include-cols`)
exclude_cols <- parse_csv_list(opt$`exclude-cols`)

base_df <- read_table(opt$base)
overlay_df <- read_table(opt$overlay)

if (opt$`normalize-cols`) {
  base_df <- normalize_columns(base_df)
  overlay_df <- normalize_columns(overlay_df)
  keys <- gsub(" +", "_", tolower(trimws(keys)))
  include_cols <- gsub(" +", "_", tolower(trimws(include_cols)))
  exclude_cols <- gsub(" +", "_", tolower(trimws(exclude_cols)))
}

check_keys(base_df, keys, "base")
check_keys(overlay_df, keys, "overlay")

base_non_keys <- setdiff(names(base_df), keys)
if (length(include_cols) > 0) {
  overlay_cols <- setdiff(intersect(include_cols, names(overlay_df)), keys)
} else {
  overlay_cols <- setdiff(names(overlay_df), keys)
}
if (length(exclude_cols) > 0) {
  overlay_cols <- setdiff(overlay_cols, exclude_cols)
}
if (length(overlay_cols) == 0) {
  stop("Tidak ada kolom overlay yang dipilih. Cek include/exclude.")
}

overlay_uniq <- overlay_df %>%
  select(all_of(c(keys, overlay_cols))) %>%
  distinct(across(all_of(keys)), .keep_all = TRUE)

join_fn <- switch(
  opt$how,
  left = left_join,
  inner = inner_join,
  right = right_join,
  full = full_join,
  stop("--how harus salah satu: left|inner|right|full")
)

merged <- join_fn(base_df, overlay_uniq, by = keys, suffix = c("_base", "_ovr"))

overlap_cols <- intersect(overlay_cols, base_non_keys)

for (col in overlap_cols) {
  c_base <- paste0(col, "_base")
  c_ovr <- paste0(col, "_ovr")

  if (opt$method == "coalesce") {
    merged[[col]] <- dplyr::coalesce(merged[[c_base]], merged[[c_ovr]])
  } else if (opt$method == "replace") {
    merged[[col]] <- dplyr::coalesce(merged[[c_ovr]], merged[[c_base]])
  } else if (opt$method == "keep_base") {
    merged[[col]] <- merged[[c_base]]
  } else if (opt$method == "keep_overlay") {
    merged[[col]] <- merged[[c_ovr]]
  } else {
    stop("--method harus salah satu: coalesce|replace|keep_base|keep_overlay")
  }
}

# Rename kolom non-overlap yang kena suffix
for (n in names(merged)) {
  if (grepl("_base$", n)) {
    original <- sub("_base$", "", n)
    if (!(original %in% overlap_cols)) {
      names(merged)[names(merged) == n] <- original
    }
  }
}
for (n in names(merged)) {
  if (grepl("_ovr$", n)) {
    original <- sub("_ovr$", "", n)
    if (!(original %in% overlap_cols) && !(original %in% names(base_df))) {
      names(merged)[names(merged) == n] <- original
    }
  }
}

# Drop kolom suffix overlap
drop_cols <- c(paste0(overlap_cols, "_base"), paste0(overlap_cols, "_ovr"))
merged <- merged %>% select(-any_of(drop_cols))

matched_n <- nrow(semi_join(base_df, overlay_uniq, by = keys))
unmatched_base_n <- nrow(anti_join(base_df, overlay_uniq, by = keys))
unmatched_overlay_n <- nrow(anti_join(overlay_uniq, base_df, by = keys))

report <- list(
  base_rows = nrow(base_df),
  overlay_rows = nrow(overlay_df),
  overlay_rows_after_dedup = nrow(overlay_uniq),
  output_rows = nrow(merged),
  matched_in_base = matched_n,
  unmatched_in_base = unmatched_base_n,
  unmatched_in_overlay = unmatched_overlay_n,
  keys = keys,
  how = opt$how,
  method = opt$method,
  overlay_columns_used = overlay_cols,
  overlap_columns = overlap_cols
)

write_table(merged, opt$output)

if (!is.null(opt$report) && nzchar(opt$report)) {
  dir.create(dirname(opt$report), recursive = TRUE, showWarnings = FALSE)
  write(toJSON(report, auto_unbox = TRUE, pretty = TRUE), file = opt$report)
}

cat("[OK] Overlay selesai\n")
cat(toJSON(report, auto_unbox = TRUE, pretty = TRUE), "\n")
